### ISSUE TITLE
Design Comprehensive Question Paper System for ML Course Assessment

### ISSUE THREAD

**Issue by eltyagi** [Lakshya3190/Cheatsheets#1]
# Question Paper System Design for Homemade Machine Learning Course

## Overview
Create a comprehensive assessment system with multiple question papers for internal tests and semester exams, featuring proper versioning and varied difficulty levels.

### 1. Question Paper Structure
- **Internal Tests**: 30-45 minutes, focused on specific topics
- **Semester Exams**: 90-120 minutes, comprehensive coverage
- **Multiple versions** per exam type to prevent cheating
- **Difficulty progression**: Beginner → Intermediate → Advanced

### 2. Version Control Strategy
```
Format: [EXAM_TYPE]_[TOPIC]_[DIFFICULTY]_v[VERSION]_[SEMESTER]_[YEAR]
Examples:
- INTERNAL_LINEAR_REG_BEGINNER_v1_FALL_2024
- SEMESTER_COMPREHENSIVE_ADVANCED_v2_SPRING_2025
```

### 3. Question Categories
- **Theory (40%)**: Concepts, algorithms, mathematical foundations
- **Code Analysis (30%)**: Understanding provided code snippets
- **Practical Implementation (20%)**: Writing/modifying code
- **Problem Solving (10%)**: Real-world application scenarios

#### Internal Test 1: Linear Regression Fundamentals
**Duration**: 45 minutes | **Difficulty**: Beginner | **Version**: v1.0

**Section A: Theory (15 marks)**
1. Define univariate linear regression and explain its use cases (3 marks)
2. What is a cost function? Explain MSE in the context of linear regression (4 marks)
3. List and explain 3 key hyperparameters in the provided LinearRegression class (4 marks)
4. Why do we split data into training and testing sets? (4 marks)

**Section B: Code Analysis (20 marks)**
```python
# Analyze this code snippet from the notebook
linear_regression = LinearRegression(x_train, y_train, polynomial_degree, sinusoid_degree)
(theta, cost_history) = linear_regression.train(learning_rate, regularization_param, num_iterations)
```
5. Explain what each parameter does in the LinearRegression constructor (5 marks)
6. What does the train() method return and why? (5 marks)
7. How would you modify this code to increase model complexity? (5 marks)
8. Identify potential issues if learning_rate is too high or too low (5 marks)

**Section C: Practical (15 marks)**
9. Write Python code to load the happiness dataset and display basic statistics (5 marks)
10. Create a scatter plot showing the relationship between GDP and Happiness Score (5 marks)
11. Calculate and display the cost function value for given predictions (5 marks)

#### Internal Test 2: Gradient Descent & Optimization
**Duration**: 45 minutes | **Difficulty**: Intermediate | **Version**: v1.0

**Section A: Theory (15 marks)**
1. Explain gradient descent algorithm and its role in linear regression (5 marks)
2. What happens when the cost function increases during training? (3 marks)
3. Compare polynomial features vs sinusoidal features for curve fitting (4 marks)
4. Define regularization and its purpose in machine learning (3 marks)

**Section B: Code Analysis (20 marks)**
```python
# From the training loop
for i in range(num_iterations):
    predictions = hypothesis(X, theta)
    cost = cost_function(predictions, y, theta, regularization_param)
    theta = theta - learning_rate * gradient
```
5. Trace through one iteration of this training loop (8 marks)
6. How does regularization_param affect the theta updates? (6 marks)
7. Suggest improvements to make this training more robust (6 marks)

**Section C: Problem Solving (15 marks)**
8. Given a scenario where training cost decreases but test cost increases, diagnose the problem and suggest solutions (8 marks)
9. Design a validation strategy for hyperparameter tuning (7 marks)

#### Semester Exam: Comprehensive Machine Learning Assessment
**Duration**: 120 minutes | **Difficulty**: Intermediate-Advanced | **Version**: v1.0

**Section A: Comprehensive Theory (30 marks)**
1. **Linear Regression Deep Dive (10 marks)**
   - Mathematical derivation of normal equation (4 marks)
   - Compare gradient descent vs normal equation approaches (3 marks)
   - Explain bias-variance tradeoff in polynomial regression (3 marks)

2. **Feature Engineering (10 marks)**
   - Design polynomial features for a 2D input space (4 marks)
   - Explain feature scaling and its importance (3 marks)
   - Discuss sinusoidal features and their applications (3 marks)

3. **Model Evaluation (10 marks)**
   - Cross-validation techniques and their benefits (4 marks)
   - Metrics beyond MSE for regression problems (3 marks)
   - Overfitting detection and prevention strategies (3 marks)

**Section B: Advanced Code Analysis (25 marks)**
4. **Repository Architecture Analysis (12 marks)**
   ```python
   # Analyze the LinearRegression class structure
   class LinearRegression:
       def __init__(self, data, labels, polynomial_degree=0, sinusoid_degree=0):
           # Implementation details
       
       def train(self, learning_rate, lambda_reg, num_iterations):
           # Training implementation
       
       def predict(self, data):
           # Prediction implementation
   ```
   - Evaluate the class design and suggest improvements (6 marks)
   - Identify potential performance bottlenecks (3 marks)
   - Propose unit testing strategy for this class (3 marks)

5. **Algorithm Implementation (13 marks)**
   - Implement gradient computation for regularized linear regression (8 marks)
   - Add early stopping mechanism to prevent overfitting (5 marks)

**Section C: Practical Implementation (25 marks)**
6. **Data Pipeline Development (15 marks)**
   - Create a robust data preprocessing pipeline (7 marks)
   - Implement train/validation/test split with proper randomization (4 marks)
   - Add data validation and error handling (4 marks)

7. **Model Comparison Framework (10 marks)**
   - Design a system to compare different polynomial degrees (5 marks)
   - Implement automated hyperparameter grid search (5 marks)

**Section D: Real-World Application (20 marks)**
8. **Business Problem Solving (20 marks)**
   - Given a new dataset (e.g., house prices), design a complete ML pipeline (10 marks)
   - Discuss potential limitations and mitigation strategies (5 marks)
   - Present results to non-technical stakeholders (5 marks)

#### Final Exam: Machine Learning Engineering Excellence
**Duration**: 150 minutes | **Difficulty**: Advanced | **Version**: v1.0

**Section A: Theoretical Mastery (25 marks)**
1. **Mathematical Foundations (10 marks)**
   - Derive the gradient descent update rule from first principles (5 marks)
   - Prove convergence conditions for gradient descent (5 marks)

2. **Advanced Concepts (15 marks)**
   - Compare different regularization techniques (L1, L2, Elastic Net) (5 marks)
   - Analyze computational complexity of different approaches (5 marks)
   - Discuss distributed training considerations (5 marks)

**Section B: System Design (30 marks)**
3. **Scalable ML Pipeline (20 marks)**
   - Design a production-ready ML service architecture (10 marks)
   - Implement model versioning and rollback mechanisms (5 marks)
   - Add monitoring and alerting for model performance (5 marks)

4. **Code Quality & Testing (10 marks)**
   - Write comprehensive unit tests for the LinearRegression class (5 marks)
   - Implement integration tests for the complete pipeline (5 marks)

**Section C: Advanced Implementation (25 marks)**
5. **Performance Optimization (15 marks)**
   - Vectorize the current implementation for better performance (8 marks)
   - Add support for mini-batch gradient descent (7 marks)

6. **Extension Development (10 marks)**
   - Implement feature selection algorithms (5 marks)
   - Add support for different cost functions (5 marks)

**Section D: Research & Innovation (20 marks)**
7. **Literature Review (10 marks)**
   - Compare the repository's approach with modern ML frameworks (5 marks)
   - Suggest improvements based on recent research (5 marks)

8. **Original Implementation (10 marks)**
   - Implement a novel feature engineering technique (10 marks)

### File Naming Convention
```
question_papers/
├── internal_tests/
│   ├── v1.0/
│   │   ├── INTERNAL_LINEAR_REG_BEGINNER_v1_FALL_2024.pdf
│   │   ├── INTERNAL_GRADIENT_DESC_INTERMEDIATE_v1_FALL_2024.pdf
│   │   └── solutions/
│   ├── v1.1/
│   └── v2.0/
├── semester_exams/
│   ├── v1.0/
│   │   ├── SEMESTER_COMPREHENSIVE_INTERMEDIATE_v1_FALL_2024.pdf
│   │   └── solutions/
│   └── v2.0/
└── final_exams/
    ├── v1.0/
    └── v2.0/
```

### Change Log Template
```markdown
# Question Paper Change Log

## Version 1.1 (2024-08-15)
### Added
- New code analysis questions for regularization
- Practical implementation of cross-validation

### Changed
- Updated difficulty level for gradient descent questions
- Improved clarity in problem statements

### Removed
- Deprecated questions about outdated practices

## Version 1.0 (2024-07-07)
### Added
- Initial release of comprehensive question bank
- Beginner, Intermediate, and Advanced difficulty levels
- Complete solution guides
```

### Scoring Guidelines
- **Excellent (90-100%)**: Complete understanding, optimal solutions, excellent code quality
- **Good (75-89%)**: Good understanding, working solutions, minor issues
- **Satisfactory (60-74%)**: Basic understanding, functional solutions, some errors
- **Needs Improvement (<60%)**: Incomplete understanding, major errors

### Code Quality Criteria
- Correctness and functionality
- Code readability and documentation
- Proper error handling
- Efficient implementation
- Following Python best practices

## Implementation Timeline
1. **Week 1**: Develop beginner-level internal tests
2. **Week 2**: Create intermediate-level assessments
3. **Week 3**: Design advanced semester exams
4. **Week 4**: Review, refine, and create solution guides
5. **Week 5**: Set up version control system and documentation

## Success Metrics
- Student performance improvement over semester
- Reduced academic dishonesty through versioning
- Comprehensive skill assessment across difficulty levels
- Preparation for real-world ML engineering challenges

This system provides a robust foundation for assessing student understanding while maintaining academic integrity through proper versioning and varied difficulty levels.
